### PR TITLE
Localizer based labeling

### PR DESCRIPTION
**Pull request by yw7**
Currently, the iterative_label.py script does not support images without any of the landmarks (C2, T1, T12, Sacrum).
To solve this issue, we can update the iterative_label.py script to get localizer segmentation as an input to initialize labels for discs and vertebrae. The localizer have to be in same space, but no need for exact match, only that most of the voxels in the first vertebrae and disc will be overlapping with the provided image. The script acts as follows:
1. Transform localizer into the image space.
1. For upper vertebrae and upper disc in the localizer:
   1. Get the superior disc/vertebra label from the matched disc/vertebra in localizer.
   1. Loop over the rest of the discs/vertebrae in the input image and map them sequentially as usual.

Instructions for new localizer based labeling:

https://github.com/neuropoly/totalspineseg/blob/07069b9bc1386be46713f71d709898f7631f8c42/README.md#L51-L86

checkout this branch:
```bash
cd totalspineseg && git checkout localizer-based-labeling && cd ..
```

https://github.com/neuropoly/totalspineseg/blob/07069b9bc1386be46713f71d709898f7631f8c42/README.md#L175-L209

Localizer

### CODE REVIEW

**Comment by yw7** [comment]
Sould we make a new version here, and create new tag before merging?
https://github.com/neuropoly/totalspineseg/blob/c5685056efb66bdd2bc55e58919d379064d2b434/pyproject.toml#L3

File: totalspineseg/utils/iterative_label.py [hunk 1]
```diff
@@ -34,6 +36,10 @@ def main():
         '--output-segs-dir', '-o', type=Path, required=True,
         help='Folder to save output segmentations.'
     )
+    parser.add_argument(
+        '--locs-dir', '-l', type=Path, default=None,
```
Comment (valosekj) [inline totalspineseg/utils/iterative_label.py:40]: maybe add `required=False,` as this is an optional flag

**Review by jcohenadad — COMMENTED** [review]
There is a lot of changes in this PR, so it is impossible for me to review all of them exhaustively. I could try the new segmentation method on data, but for that I would need instructions on how to test this new segmentation method (what command to run, etc.)

**Comment by yw7** [comment]
> There is a lot of changes in this PR, so it is impossible for me to review all of them exhaustively. I could try the new segmentation method on data, but for that I would need instructions on how to test this new segmentation method (what command to run, etc.)

Thank you for reviewing the PR!
I've aded instructions in the readme here:

https://github.com/neuropoly/totalspineseg/tree/localizer-based-labeling?tab=readme-ov-file#localizer-based-labeling

Regarding the changes, is it possible to remove from the view here the commits that are already merged to main? I see that they include here also

File: totalspineseg/utils/iterative_label.py (hunk 1 above)
Comment (yw7) [reply totalspineseg/utils/iterative_label.py:40]: Thanks a lot for reviewing the PR and the comment,
Are you sure it needed? Because for what I understand the default is False for non positional arguments.

https://docs.python.org/3/library/argparse.html#required

**Review by jcohenadad — CHANGES_REQUESTED** [review]
Issue:

~~~
julien-macbook:~/code/totalspineseg $ python3 -m venv venv
julien-macbook:~/code/totalspineseg $ source venv/bin/activate 
(venv) julien-macbook:~/code/totalspineseg $ python3 -m pip install -e totalspineseg 
Obtaining file:///Users/julien/code/totalspineseg/totalspineseg
ERROR: file:///Users/julien/code/totalspineseg/totalspineseg does not appear to be a Python project: neither 'setup.py' nor 'pyproject.toml' found.

[notice] A new release of pip is available: 23.0.1 -> 24.2
[notice] To update, run: pip install --upgrade pip
(venv) julien-macbook:~/code/totalspineseg $ pip install --upgrade pip
Requirement already satisfied: pip in ./venv/lib/python3.10/site-packages (23.0.1)
Collecting pip
  Downloading pip-24.2-py3-none-any.whl (1.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 33.7 MB/s eta 0:00:00
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 23.0.1
    Uninstalling pip-23.0.1:
      Successfully uninstalled pip-23.0.1
Successfully installed pip-24.2
(venv) julien-macbook:~/code/totalspineseg $ python3 -m pip install -e totalspineseg 
Obtaining file:///Users/julien/code/totalspineseg/totalspineseg
ERROR: file:///Users/julien/code/totalspineseg/totalspineseg does not appear to be a Python project: neither 'setup.py' nor 'pyproject.toml' found.
~~~

**Comment by jcohenadad** [comment]
image

is this really necessary? can't we have a CLI that just points to the image to run the inference on?

**Comment by yw7** [comment]
> Issue:
> 
> ```
> julien-macbook:~/code/totalspineseg $ python3 -m venv venv
> julien-macbook:~/code/totalspineseg $ source venv/bin/activate 
> (venv) julien-macbook:~/code/totalspineseg $ python3 -m pip install -e totalspineseg 
> Obtaining file:///Users/julien/code/totalspineseg/totalspineseg
> ERROR: file:///Users/julien/code/totalspineseg/totalspineseg does not appear to be a Python project: neither 'setup.py' nor 'pyproject.toml' found.
> 
> [notice] A new release of pip is available: 23.0.1 -> 24.2
> [notice] To update, run: pip install --upgrade pip
> (venv) julien-macbook:~/code/totalspineseg $ pip install --upgrade pip
> Requirement already satisfied: pip in ./venv/lib/python3.10/site-packages (23.0.1)
> Collecting pip
>   Downloading pip-24.2-py3-none-any.whl (1.8 MB)
>      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 33.7 MB/s eta 0:00:00
> Installing collected packages: pip
>   Attempting uninstall: pip
>     Found existing installation: pip 23.0.1
>     Uninstalling pip-23.0.1:
>       Successfully uninstalled pip-23.0.1
> Successfully installed pip-24.2
> (venv) julien-macbook:~/code/totalspineseg $ python3 -m pip install -e totalspineseg 
> Obtaining file:///Users/julien/code/totalspineseg/totalspineseg
> ERROR: file:///Users/julien/code/totalspineseg/totalspineseg does not appear to be a Python project: neither 'setup.py' nor 'pyproject.toml' found.
> 
> (venv) julien-macbook:~/code/totalspineseg $ ll
> total 40
> drwxr-xr-x   9 julien  staff    288 10 Sep 16:52 .
> drwxr-xr-x  43 julien  staff   1376 10 Sep 16:52 ..
> drwxr-xr-x  12 julien  staff    384 10 Sep 16:52 .git
> -rw-r--r--   1 julien  staff   3316 10 Sep 16:52 .gitignore
> -rw-r--r--   1 julien  staff  10262 10 Sep 16:52 README.md
> -rw-r--r--   1 julien  staff   3308 10 Sep 16:52 pyproject.toml
> drwxr-xr-x   6 julien  staff    192 10 Sep 16:52 scripts
> drwxr-xr-x   6 julien  staff    192 10 Sep 16:52 totalspineseg
> drwxr-xr-x   6 julien  staff    192 10 Sep 16:52 venv
> ```
> 
> correct instruction:
> 
> ```
> (venv) julien-macbook:~/code/totalspineseg $ python3 -m pip install -e .
> ```
> 
> nevermind-- i see that we were not supposed to go into the git cloned repos (which is usually common practice, so it's a bit confusing)-- i suggest to instruct to go into the repos and then install from there

Thank you!! this is actually confusing. We should make it more intuitive.
Will it be more better if we change the instructions to:
```bash
cd totalspineseg
python3 -m pip install -e .
cd ..
```

**Comment by yw7** [comment]
> is this really necessary? can't we have a CLI that just points to the image to run the inference on?
This is actually good point to discuss. The current implementation is not perfect.
We need it mainly for training so tha training script will know where the data is and where the code `resources` are. For infernce this used for constand dir to store the model pretrained weights.
For inference only you can use the `--data-dir` or `-d` argument to tell the code where the model weights are stored. But wouldn't it be more convinced to have a constant variable in the venv to save the path to the model weights?

**Comment by jcohenadad** [comment]
> This is actually good point to discuss. The current implementation is not perfect.
We need it mainly for training, so tha training script will know where the data is and where the resources are. For infernce this used for constand dir to store the model pretrained weights.
For inference only the --data-dir or -d argument available instead to tell the code where the model weights are stored. But wouldn't it be more convinced to have a constant variable in the venv to save the path to the model weights?

I wouldn't worry too much about these fine details, given that ultimately this method will be callable via SCT. But we should already think of an appropriate API to plug it inside SCT. Tagging @joshuacwnewton @mguaypaq so we can start thinking about it

**Comment by joshuacwnewton** [comment]
Thank you for the ping. I'll take a look at this first thing tomorrow and begin thinking about ways that this project can integrate with SCT. I appreciate you keeping us in the loop! :)

**Comment by NathanMolinier** [comment]
Regarding this PR I am having troubles. It seems like localizer are not being used
Screenshot 2024-09-13 at 10 29 48

I ran this command:
```
totalspineseg raw_crop_reg/MTS out_reg_loc/ --loc out_loc/step2_output/MTS --suffix _MTS --loc-suffix _T2w
```

**Comment by NathanMolinier** [comment]
Sorry for the delay but here are some results that I created to validate the relevance of the PR
Screenshot 2024-09-13 at 17 08 45

File: totalspineseg/inference.py [hunk 2]
```diff
@@ -12,22 +13,46 @@ def main():
...
+        'output', type=Path,
         help='The output folder where the model outputs will be stored.'
     )
+    parser.add_argument(
+        '--loc', '-l', type=Path, default=None,
+        help=' '.join(f'''
+            Folder containing localizers segmentations or a single .nii.gz localizer segmentation to use for detecting first vertebrae and disc if C1 and C2-C3 disc or the Sacrum and L5-S disc not found in the image, Optional.
+            This is the output of the model applied on localizer images. It can be the output of step 2, or step 1 if you only want to run step 1 (step1 flag).
+            The algorithm will transform the localizer to the segmentation space and use it to detect the matching vertebrae and discs.
+            Mathcing will based on the magority of the voxels of the first vertebrae or disc in the localizer, that intersect with image.
```
Comment (NathanMolinier) [inline totalspineseg/inference.py:44]:
```suggestion
            Matching will based on the majority of the voxels of the first vertebrae or disc in the localizer, that intersect with image.
```

File: totalspineseg/utils/iterative_label.py (hunk 1 above)
Comment (NathanMolinier) [reply totalspineseg/utils/iterative_label.py:40]: Yes, if you don't specify, the `required` argument is set to false

File: totalspineseg/utils/iterative_label.py [hunk 3]
```diff
@@ -34,6 +36,14 @@ def main():
...
+        help=' '.join(f'''
+            Folder containing localizers segmentations to use for detecting first vertebrae and disc if init label not found, Optional.
```
Comment (NathanMolinier) [inline totalspineseg/utils/iterative_label.py:42]:
```suggestion
            Folder containing localizers' segmentations to help the labeling. Optional.
```

File: totalspineseg/utils/iterative_label.py [hunk 4]
```diff
@@ -293,48 +354,60 @@ def iterative_label_mp(
...
     seg_path = Path(seg_path)
     output_seg_path = Path(output_seg_path)
+    loc_path = loc_path and Path(loc_path)
 
     # If the output image already exists and we are not overriding it, return
     if not override and output_seg_path.exists():
         return
 
-    # Load segmentation
+    # Load segmentation and localizer images
```
[the remark this answers is not in the document]
Comment (yw7) [reply totalspineseg/utils/iterative_label.py:387]: I've changed to # Load segmentation and localizers since both of them are segmentations
Is it ok?

File: totalspineseg/utils/iterative_label.py [hunk 5]
```diff
@@ -34,6 +36,14 @@ def main():
         '--output-segs-dir', '-o', type=Path, required=True,
         help='Folder to save output segmentations.'
     )
+    parser.add_argument(
+        '--locs-dir', '-l', type=Path, default=None,
+        help=' '.join(f'''
+            Folder containing localizers segmentations to use for detecting first vertebrae and disc if init label not found, Optional.
+            The algorithm will transform the localizer to the segmentation space and use it to detect the matching vertebrae and disc if the init label not found.
```
[the remark this answers is not in the document]
Comment (yw7) [reply totalspineseg/utils/iterative_label.py:43]: Fixed here: https://github.com/neuropoly/totalspineseg/pull/56/commits/8072c625849e3d40828f33493d7b356c594ca089

File: totalspineseg/utils/iterative_label.py [hunk 6]
```diff
@@ -140,23 +170,29 @@ def main():
...
     output_disc_step = args.output_disc_step
-    vertebrea_labels = args.vertebrea_labels
-    vertebrea_extra_labels = args.vertebrea_extra_labels
+    loc_disc_labels = [_ for __ in args.loc_disc_labels for _ in (__ if isinstance(__, list) else [__])]
+    vertebrae_labels = [_ for __ in args.vertebrae_labels for _ in (__ if isinstance(__, list) else [__])]
+    vertebrae_extra_labels = [_ for __ in args.vertebrae_extra_labels for _ in (__ if isinstance(__, list) else [__])]
     init_vertebrae = dict(args.init_vertebrae)
-    output_vertebrea_step = args.output_vertebrea_step
+    output_vertebrae_step = args.output_vertebrae_step
+    loc_vertebrae_labels = [_ for __ in args.loc_vertebrae_labels for _ in (__ if isinstance(__, list) else [__])]
```
[the remark this answers is not in the document]
Comment (yw7) [reply totalspineseg/utils/iterative_label.py:188]: Fixed here: https://github.com/neuropoly/totalspineseg/pull/56/commits/8072c625849e3d40828f33493d7b356c594ca089

File: totalspineseg/utils/iterative_label.py [hunk 7]
```diff
@@ -34,6 +36,14 @@ def main():
         '--output-segs-dir', '-o', type=Path, required=True,
         help='Folder to save output segmentations.'
     )
+    parser.add_argument(
+        '--locs-dir', '-l', type=Path, default=None,
+        help=' '.join(f'''
+            Folder containing localizers segmentations to use for detecting first vertebrae and disc if init label not found, Optional.
+            The algorithm will transform the localizer to the segmentation space and use it to detect the matching vertebrae and disc if the init label not found.
+            Mathcing will based on the magority of the voxels of the first vertebrae or disc in the localizer, that intersect with the input segmentation.
```
[the remark this answers is not in the document]
Comment (yw7) [reply totalspineseg/utils/iterative_label.py:44]: Fixed here: https://github.com/neuropoly/totalspineseg/pull/56/commits/8072c625849e3d40828f33493d7b356c594ca089

File: totalspineseg/utils/iterative_label.py [hunk 8]
```diff
@@ -121,6 +143,14 @@ def main():
...
         '''.split())
     )
+    parser.add_argument(
+        '--default-superior-disc', type=int, default=0,
+        help='Default superior disc label if no init label found, defaults to 0 (Raise error if init label not found).'
+    )
+    parser.add_argument(
+        '--default-superior-vertebrae', type=int, default=0,
+        help='Default superior vertebrae label if no init label found, defaults to 0 (Raise error if init label not found).'
+    )
```
[the remark this answers is not in the document]
Comment (yw7) [reply totalspineseg/utils/iterative_label.py:153]: As we've discussed, I've added comment that this is not in use in current implementation 
https://github.com/neuropoly/totalspineseg/pull/56/commits/627238ea409cf81f0af97f80dfe0ef5883b116d8

File: totalspineseg/utils/iterative_label.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (NathanMolinier) [reply totalspineseg/utils/iterative_label.py:387]: Then mine makes more sense no ? Not that problematic

[the remark this answers is not in the document]
Comment (NathanMolinier) [reply totalspineseg/utils/iterative_label.py:387]: Technically segmentation are images but we can go with your suggestion: maybe I would just remove the `s` at the end. (`# Load segmentation and localizers` --> `# Load segmentation and localizer`)

**Comment by NathanMolinier** [comment]
Last question regarding this PR, as raised by @jcohenadad, what would be the best practice regarding the installation of the package ?

Currently we ask the user to do this
```
git clone https://github.com/neuropoly/totalspineseg.git 
pip install -e totalspineseg 
```
But we were wondering if instead it would be more intuitive for the user to do this:
```
git clone https://github.com/neuropoly/totalspineseg.git
cd totalspineseg
pip install -e .
cd ..
```

> Note: After this installation we are creating a `data` folder outside of the repository and initializing variables, that's why we need to do move back with `cd ..`.

Do you have any recommandations on this @joshuacwnewton @mguaypaq ?

**Comment by yw7** [comment]
> Last question regarding this PR, as raised by @jcohenadad, what would be the best practice regarding the installation of the package ?
> 
> Currently we ask the user to do this
> 
> ```
> git clone https://github.com/neuropoly/totalspineseg.git 
> pip install -e totalspineseg 
> ```
> 
> But we were wondering if instead it would be more intuitive for the user to do this:
> 
> ```
> git clone https://github.com/neuropoly/totalspineseg.git
> cd totalspineseg
> pip install -e .
> cd ..
> ```
> 
> > Note: After this installation we are creating a `data` folder outside of the repository and initializing variables, that's why we need to do move back with `cd ..`.
> 
> Do you have any recommandations on this @joshuacwnewton @mguaypaq ?

Can we consider merge this PR and open an issue for that as this is not related to something that changed in this PR?

**Comment by joshuacwnewton** [comment]
Agree with @yw7 -- Please feel free to tag me on any new issues discussing installation/packaging. :)

**Comment by yw7** [comment]
> Agree with @yw7 -- Please feel free to tag me on any new issues discussing installation/packaging. :)

Thanks!!
Issue opened https://github.com/neuropoly/totalspineseg/issues/64

**Review by jcohenadad — APPROVED** [review]
